### DISCUSSION TITLE
Fortress on Ymir

### DISCUSSION THREAD

**Discussion by calvinalkan** [snicco/fortress#11] [Q&A]
This discussion will be a work-in-progress to discuss the optimal Fortress configuration on [Ymir](https://ymirapp.com/).

The purpose is to later merge this into a "Guide" in the [Ymir Docs](https://docs.ymirapp.com/introduction.html).

**Comment by calvinalkan** [comment]
## Configuration Cache and Log Files.

Fortress cannot build its cached configuration file on Ymir since it's read-only. 

This will throw an exception, since Fortress expected the log/cache directory to be existing and writable.

There are three ways to handle this:

###  1. Use /tmp as a log/cache directory | doesn't work

```php
$_SERVER['SNICCO_FORTRESS_CACHE_DIR'] = '/tmp'
$_SERVER['SNICCO_FORTRESS_LOG_DIR'] = '/tmp'
```

`/tmp` is the only writable directory on lambda.

**Downsides**:

- The cache is rebuilt for each new lambda instance (might not be relevant).
- Possible file name conflicts with existing files in `/tmp` (unlikely)

###  2. Use /tmp/fortress as a log/cache directory

```php
if(!is_dir('/tmp/fortress')) {
   mkdir('/tmp/fortress/cache', 0755, true);
   mkdir('/tmp/fortress/log', 0755, true);

} 

$_SERVER['SNICCO_FORTRESS_CACHE_DIR'] = '/tmp/fortress/cache'
$_SERVER['SNICCO_FORTRESS_LOG_DIR'] = '/tmp/fortress/log'
```

**Downsides**:

- The cache is rebuilt for each new lambda instance (might not be relevant).
- One additional "is_dir" check for each request (unless Ymir can bake this into the docker runtime).

### 3. Deploy cached configuration files

Lastly, it's possible to deploy a cached configuration that was built locally.

There are currently no env (db, server, filesystem) related configuration settings in Fortress; a local configuration should always be reproducible remotely.

This would mean the Fortress configuration cache must be put into version control and included in the Ymir deployment.

The build step should include the following:

```yaml
id: 1
name: project-name
type: wordpress
environments:
  environment-name:
    build:
      - 'composer install'  # Install Fortress via composer
      - 'wp snicco/fortress cache:clear && wp snicco/fortress config:test --reload-on-success' # Clear current cache & rebuilt sources
```


```php
$_SERVER['SNICCO_FORTRESS_LOG_DIR'] = '/tmp'
```


**Downsides**:

- The cache directory is directly accessible from the webroot in `wp-content/mu-plugins/snicco-fortress/var/cache/`

**Reply by calvinalkan** [reply]
I prefer option 2) with /tmp/fortress already created by Ymir.

This will need confirmation from @carlalexander to see how often a cache rebuild would be triggered when autoscaling.

**Reply by calvinalkan** [reply]
Access to `wp-content/mu-plugins/snicco-fortress/var/cache/`  can likely be prevented on the AWS level.

Or, if it's a bedrock project, both cache/logs could live in the project's root, which means they should not be web-accessible by default.
```shell
# Create a fortress directory in the bedrock project root.
mkdir -p ./fortress/cache
touch .fortress/cache/.gitkeep
```

```php
$_SERVER['SNICCO_FORTRESS_CACHE_DIR'] = dirname(ABSPATH).'/fortress/cache'

$_SERVER['SNICCO_FORTRESS_LOG_DIR'] = '/tmp'
```

This would probably not work for "normal" wp projects tho.

**Reply by carlalexander** [reply]
Yeah, I'd go with 2.) But just to be clear, if you log to `/tmp`, you won't be seeing the logs anywhere since only stuff sent to `STDERR` goes to CloudWatch. Do you log with `error_log` as well?

**Reply by calvinalkan** [reply]
Yes, right now we use daily rotating file + error_log / stderr. 

So it also goes to Cloudwatch. 

In the future we'll allow providing your own logger, so you could disable file logger entirely.

**Reply by calvinalkan** [reply]
@carlalexander 

So how often is a new lamda invoked? 

Cache build right now is single digit ms. 

But it might be more in the future, working on pretty extensive semantic validation of config right now.

**Reply by pmgarman** [reply]
Testing this - I have these env vars set, and I get an error related to the /tmp dir file ops. I am not sure why it's trying to use a .ymir dir instead of just /tmp ... but here you go!

```
SNICCO_FORTRESS_CACHE_DIR=/tmp
SNICCO_FORTRESS_LOG_DIR=/tmp
```

```
➜  garman git:(ymir) ✗ ymir wp

 Please enter the WP-CLI command to run:
 > snicco/fortress shared trigger-activation --vvv

Running "wp snicco/fortress shared trigger-activation --vvv" on "staging" environment

PHP Fatal error:  Uncaught Snicco\ScopedVendor\Webmozart\Assert\InvalidArgumentException: Could not remove directory /tmp/.ymir in /var/task/web/app/mu-plugins/snicco-fortress/vendor/webmozart/assert/src/Assert.php:1601
Stack trace:
#0 /var/task/web/app/mu-plugins/snicco-fortress/vendor/webmozart/assert/src/Assert.php(496): Snicco\ScopedVendor\Webmozart\Assert\Assert::reportInvalidArgument()
#1 /var/task/web/app/mu-plugins/snicco-fortress/src/Shared/Infrastructure/Configuration/PurgeFileCache.php(57): Snicco\ScopedVendor\Webmozart\Assert\Assert::true()
#2 /var/task/web/app/mu-plugins/snicco-fortress/src/Shared/Infrastructure/Configuration/PurgeFileCache.php(34): Snicco\Enterprise\Fortress\Shared\Infrastructure\Configuration\PurgeFileCache->removeDirectoryContentsRecursive()
#3 /var/task/web/app/mu-plugins/snicco-fortress/boot/FortressKernelFactory.php(76): Snicco\Enterprise\Fortress\Shared\Infrastructure\Configuration\PurgeFileCache->__invoke()
#4 /var/task/web/app/mu-plugins/snicco-fortress/boot/FortressKernelFactory.php(57): Snicco\Enterprise\Fortress\FortressKernelFactory->maybeInvalidateCache()
#5 /var/task/web/app/mu-plugins/snicco-fortress/boot/create-kernel.php(9): Snicco\Enterprise\Fortress\FortressKernelFactory->create()
#6 /var/task/web/app/mu-plugins/snicco-fortress/main.php(66): WP_CLI\Runner::Snicco\Enterprise\Fortress\{closure}()
#7 /var/task/web/app/mu-plugins/snicco-fortress/main.php(133): WP_CLI\Runner->Snicco\Enterprise\Fortress\{closure}()
#8 /var/task/web/app/mu-plugins/fortress.php(17): require_once('...')
#9 /var/task/web/wp/wp-settings.php(387): include_once('...')
#10 phar:///var/task/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1271): require('...')
#11 phar:///var/task/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1192): WP_CLI\Runner->load_wordpress()
#12 phar:///var/task/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Bootstrap/LaunchRunner.php(23): WP_CLI\Runner->start()
#13 phar:///var/task/bin/wp/vendor/wp-cli/wp-cli/php/bootstrap.php(77): WP_CLI\Bootstrap\LaunchRunner->process()
#14 phar:///var/task/bin/wp/vendor/wp-cli/wp-cli/php/wp-cli.php(27): WP_CLI\bootstrap()
#15 phar:///var/task/bin/wp/php/boot-phar.php(11): include('...')
#16 /var/task/bin/wp(4): include('...')
#17 {main}
  thrown in /var/task/web/app/mu-plugins/snicco-fortress/vendor/webmozart/assert/src/Assert.php on line 1601
Fatal error: Uncaught Snicco\ScopedVendor\Webmozart\Assert\InvalidArgumentException: Could not remove directory /tmp/.ymir in /var/task/web/app/mu-plugins/snicco-fortress/vendor/webmozart/assert/src/Assert.php:1601
Stack trace:
#0 /var/task/web/app/mu-plugins/snicco-fortress/vendor/webmozart/assert/src/Assert.php(496): Snicco\ScopedVendor\Webmozart\Assert\Assert::reportInvalidArgument()
#1 /var/task/web/app/mu-plugins/snicco-fortress/src/Shared/Infrastructure/Configuration/PurgeFileCache.php(57): Snicco\ScopedVendor\Webmozart\Assert\Assert::true()
#2 /var/task/web/app/mu-plugins/snicco-fortress/src/Shared/Infrastructure/Configuration/PurgeFileCache.php(34): Snicco\Enterprise\Fortress\Shared\Infrastructure\Configuration\PurgeFileCache->removeDirectoryContentsRecursive()
#3 /var/task/web/app/mu-plugins/snicco-fortress/boot/FortressKernelFactory.php(76): Snicco\Enterprise\Fortress\Shared\Infrastructure\Configuration\PurgeFileCache->__invoke()
#4 /var/task/web/app/mu-plugins/snicco-fortress/boot/FortressKernelFactory.php(57): Snicco\Enterprise\Fortress\FortressKernelFactory->maybeInvalidateCache()
#5 /var/task/web/app/mu-plugins/snicco-fortress/boot/create-kernel.php(9): Snicco\Enterprise\Fortress\FortressKernelFactory->create()
#6 /var/task/web/app/mu-plugins/snicco-fortress/main.php(66): WP_CLI\Runner::Snicco\Enterprise\Fortress\{closure}()
#7 /var/task/web/app/mu-plugins/snicco-fortress/main.php(133): WP_CLI\Runner->Snicco\Enterprise\Fortress\{closure}()
#8 /var/task/web/app/mu-plugins/fortress.php(17): require_once('...')
#9 /var/task/web/wp/wp-settings.php(387): include_once('...')
#10 phar:///var/task/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1271): require('...')
#11 phar:///var/task/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1192): WP_CLI\Runner->load_wordpress()
#12 phar:///var/task/bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Bootstrap/LaunchRunner.php(23): WP_CLI\Runner->start()
#13 phar:///var/task/bin/wp/vendor/wp-cli/wp-cli/php/bootstrap.php(77): WP_CLI\Bootstrap\LaunchRunner->process()
#14 phar:///var/task/bin/wp/vendor/wp-cli/wp-cli/php/wp-cli.php(27): WP_CLI\bootstrap()
#15 phar:///var/task/bin/wp/php/boot-phar.php(11): include('...')
#16 /var/task/bin/wp(4): include('...')
#17 {main}
  thrown in /var/task/web/app/mu-plugins/snicco-fortress/vendor/webmozart/assert/src/Assert.php on line 1601
Error: There has been a critical error on this website.Learn more about troubleshooting WordPress. There has been a critical error on this website.
```

**Reply by calvinalkan** [reply]
@pmgarman 

You're right. 

The entire Cache directory needs to be reserved for fortress stuff only. I forgot that. We recursively purge the cache directory, which obviously doesnt work.

Should be option2:

tmp/fortress/cache

And

tmp/fortress/log

**Reply by pmgarman** [reply]
This configuration worked. Cost only 48microseconds ;) 10x slower without the isdir checks for what its worth.

Env
```
SNICCO_FORTRESS_CACHE_DIR=/tmp/fortress/cache
SNICCO_FORTRESS_LOG_DIR=/tmp/fortress/log
```

In wp-config, right before 
```php
if(!is_dir(Config::get('SNICCO_FORTRESS_CACHE_DIR'))) {
	mkdir(Config::get('SNICCO_FORTRESS_CACHE_DIR'), 0755, true);
}

if(!is_dir(Config::get('SNICCO_FORTRESS_LOG_DIR'))) {
	mkdir(Config::get('SNICCO_FORTRESS_LOG_DIR'), 0755, true);
}
```

**Reply by calvinalkan** [reply]
> 10x slower without the isdir checks for what its worth.

I'm confused. Do you mean 10x slower WITH the is_dir() ?

Can also change it do:

```php

@mkdir(Config::get('SNICCO_FORTRESS_CACHE_DIR'), 0755, true);
@mkdir(Config::get('SNICCO_FORTRESS_LOG_DIR'), 0755, true);
```

But yeah, ideally the ymir runtime would pre-create these directories in the docker image.

**Reply by pmgarman** [reply]
This code is 48 microseconds

```php
if(!is_dir(Config::get('SNICCO_FORTRESS_CACHE_DIR'))) {
	mkdir(Config::get('SNICCO_FORTRESS_CACHE_DIR'), 0755, true);
}

if(!is_dir(Config::get('SNICCO_FORTRESS_LOG_DIR'))) {
	mkdir(Config::get('SNICCO_FORTRESS_LOG_DIR'), 0755, true);
}
```

This code is 480microseconds

```php
mkdir(Config::get('SNICCO_FORTRESS_CACHE_DIR'), 0755, true);
mkdir(Config::get('SNICCO_FORTRESS_LOG_DIR'), 0755, true);
```

`mkdir` will succeed even if the directory already exists, so I figured I'd check if it was any more efficient to just let it roll, but it's a lot slower.

**Reply by pmgarman** [reply]
Re: the docker image. The ymir runtime just uploads the application ZIP, I don't know that it _can_ create directories in /tmp on boot. Though it's possible for sites to move to a container on Lambda via Ymir... in which case they absolutely can do anything they want.

**Reply by calvinalkan** [reply]
> The ymir runtime just uploads the application ZIP, I don't know that it can create directories in /tmp on boot.

I think Carl could add that upstream in Ymir. Might be wrong. @carlalexander 

> so I figured I'd check if it was any more efficient to just let it roll, but it's a lot slower.

That's odd, but hey.

I think what's more relevant is how long it takes to "build" Fortress's cached configuration(s). 

Right now, it's very fast because we only perform syntax validation, but I'm working on deep semantic validation of all values right now, which is built with the assumption that the configuration is NOT rebuilt on every request. 

Before we go down this path further, @carlalexander Can you give insight on how requests/second correlate to "clean" lambda invocations?

How often would we get a clean filesystem?

**Reply by calvinalkan** [reply]
Worst case, I can add an abstraction for the cached config storage/retrieval, allowing either Filesytem or Redis.

**Reply by carlalexander** [reply]
@pmgarman You should use container image as a default. It's just flat out better. I don't make default because:

1) Didn't exist at first so to maintain BC I keep zip default
2) WordPress devs and Docker 🫠

@calvinalkan It really depends. It's not a request per second thing. It's the maximum # of invocation per lambda. So could take you a minute or two to hit it or a hour or never because the lambda gets destroyed. By default, the maximum # of invocations is 100. But it can be changed via an [environment variable](https://github.com/ymirapp/php-runtime/blob/main/src/Runtime.php#L103). This is also useful with Relay since you lose the in-memory cache whenever the lambda gets destroyed. So I had a customer ask to be able to change it.

**Comment by calvinalkan** [comment]
## Fortress Secrets

Fortress's Secrets should be [Ymir ENV secrets](https://docs.ymirapp.com/projects/environments.html#secrets). 

Fortress secrets can be generated to stdout by running:

`php /path-to-fortress-locally/bin/prod/generate-secrets.php`

The output of this command can then be piped (with some xargs magic) to `ymir environment:secret:change`

Secrets must be generated per ymir environment **before** Fortress is deployed.

Secrets must never be changed after creating them in one env.

**Reply by calvinalkan** [reply]
This part could potentially be integrated into Ymir CLI.

**Reply by carlalexander** [reply]
Yeah, I'd have to see, but it could be for sure.

**Reply by calvinalkan** [reply]
This is how secrets can be generated, it's pretty easy. 

https://github.com/snicco/fortress/blob/beta/docs/getting-started/02_preparation.md#generating-secrets-before-activation

Output is lines with this Format:

ENVVAR_NAME:VALUE

**Reply by calvinalkan** [reply]
@carlalexander 

Also, I don't know how people currently use different Ymir ENVs regarding staging => production workflows.

But if there ever is a scenario where somebody pushes staging DB => prod DB, the secrets must be the same on both sites. 

Otherwise 💣

**Reply by carlalexander** [reply]
That's a more challenging issue because secrets are kinda one way. I can't get them back for you. They'd probably need to keep copies somewhere safe (e.g. 1password) to make sure they can add them to new environments if they create some after the fact.

**Reply by calvinalkan** [reply]
@carlalexander 

Right, either store them somewhere safe or generate them once for all environments you forsee using.

That said, is that workflow even a thing on Ymir?

Pushing staging DB to prod DB?

**Reply by carlalexander** [reply]
Not right now, but it will one day 🥹

**Reply by calvinalkan** [reply]
Sounds like a problem for future carl and calvin. 

I can always create a secrets:list CLI command if needed.

**Comment by pmgarman** [comment]
## Static Assets

When ymir deploys your site, it by default strips out any non php assets, including JSON files. 

```
NOTICE: PHP message: PHP Fatal error:  Uncaught RuntimeException: Cant read file contents of file [/var/task/web/app/mu-plugins/snicco-fortress/vendor/snicco/psr7-error-handler/resources/en_US.error.json] in /var/task/web/app/mu-plugins/snicco-fortress/vendor/snicco/psr7-error-handler/src/Information/InformationProviderWithTransformation.php:49
```

In the ymir.yml adding something like this seems to resolve it. There are multiple json files in Fortress that are required to exist for the plugin to function. Examples, the JSON file from above, as well as the manifest file `snicco-fortress/public/manifest.json` - I'm not sure all the files, some come from vendors, so just requiring the whole plugin while a sledge hammer approach, likely solves for all cases...

```yaml
environments:
  staging:
    build:
      include:
        - web/app/mu-plugins/snicco-fortress
```

**Reply by calvinalkan** [reply]
Yea, there are some non PHP files in Fortress, and some in vendor.

**Reply by calvinalkan** [reply]
Also, make sure you include your custom configuration file.
